### DISCUSSION TITLE
Where to uncompress a POST request's body if not in a middleware?

### DISCUSSION THREAD

**Discussion by JivanRoquet** [fastapi/fastapi#6744] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

The example below makes all queries freeze, and I already know that this is due to the request's body being consumed within the middleware, therefore making it unavailable for what comes after `call_next()`.

Is there a better solution to handle the problem of uncompressing a compressed request body?

```Python
from fastapi import FastAPI, Request
from pydantic import BaseModel

app = FastAPI()

class MySchema(BaseModel):
    name: str

@app.middleware("http")
async def decompress_if_gzip(request: Request, call_next):
    if request.headers.get("content-encoding", "") == "gzip":
        # the request's body is compressed, so we need to decompress it
        body = await request.body()
        dec = gzip.decompress(body)
        request._body = dec # <-- if only things were that easy
    response = await call_next(request)
    return response

@app.post("/hello")
def hello(query: MySchema):
    return [f"Hello, {query.name}"]
```

### Description

I would like to send POST requests with a compressed body to FastAPI. I haven't found any ready-made middleware that enables this, and I'm actually wondering if a middleware is the best place to do so.

This issue https://github.com/tiangolo/fastapi/issues/394 mentions that it's a bad idea to consume request data in the middleware. Well, yes, but then... where else?

I can't do it in the endpoint method (`def hello(query: MySchema)`) because by then, all validation has already happened and the compressed body has (rightly so) been rejected by the validation process.

Any of these two alternatives would be interesting:

- either a different place to decompress the body than the middleware
- or a way to make the decompression work in the middleware (and the query not freeze anymore)

### Environment

* OS: Linux/MacOS:
* FastAPI Version 0.63.0
* Python version: 3.9.2

**Comment by waynerv** [comment, marked as answer]
This issue #394 already give 3 solutions:
1. use a global dependency
2. use a custom Requtest or Router
3. still in middleware

**Comment by JivanRoquet** [comment]
I had already read issue #394 but it doesn't really give a specific solution for this use case.

This article in the "advanced" documentation, however, does it perfectly: https://fastapi.tiangolo.com/advanced/custom-request-and-route/#accessing-the-request-body-in-an-exception-handler

**Comment by tiangolo** [comment]
Thanks for the help here @waynerv ! 👏  🙇 

Thanks for reporting back and closing the issue @JivanRoquet  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
